libraries/chain/controller.cpp

```diff
@@ -4312,8 +4312,17 @@ struct controller_impl {
          if (qc) {
             verify_qc_future.get();
          }
-         integrate_received_qc_to_block(bsp); // Save the received QC as soon as possible, no matter whether the block itself is valid or not
-         consider_voting(bsp, use_thread_pool_t::no);
+         if (async_voting == async_t::yes) {
+            boost::asio::post(thread_pool.get_executor(), [this, bsp=bsp]() {
+               try {
+                  integrate_received_qc_to_block(bsp); // Save the received QC as soon as possible, no matter whether the block itself is valid or not
```

Comment:
Should `integrate_received_qc_to_block(bsp)` be done outside this thread? because the condition above is only about `(async_voting == async_t::yes)`. It would be safer to only put `consider_voting()` in the thread pool.

Reply:
The `integrate_received_qc_to_block` seems rather expensive:
- Fetches a block from the forkdb by number (mutex, multi-index search (multiple lookups), shared_ptr copy)
- Set qc in block (at least one mutex, sometimes two)
- Update FSI (mutex and a write to a file with a file flush)

Would like to not have to wait on all of that before a block can be applied.